coderd/notifications/dispatch/smtp.go

```diff
@@ -0,0 +1,338 @@
+
+	//go:embed smtp/html.gotmpl
+	htmlTemplate string
+	//go:embed smtp/plaintext.gotmpl
+	plainTemplate string
+)
+
+// SMTPHandler is responsible for dispatching notification messages via SMTP.
+// NOTE: auth and TLS is currently *not* enabled in this initial thin slice.
+// TODO: implement auth
+// TODO: implement TLS
+type SMTPHandler struct {
+	cfg codersdk.NotificationsEmailConfig
+	log slog.Logger
+}
+
+func NewSMTPHandler(cfg codersdk.NotificationsEmailConfig, log slog.Logger) *SMTPHandler {
+	return &SMTPHandler{cfg: cfg, log: log}
+}
+
+func (*SMTPHandler) NotificationMethod() database.NotificationMethod {
+	return database.NotificationMethodSmtp
+}
+
+func (s *SMTPHandler) Dispatcher(payload types.MessagePayload, titleTmpl, bodyTmpl string) (DeliveryFunc, error) {
+	// First render the subject & body into their own discrete strings.
+	subject, err := render.Plaintext(titleTmpl)
+	if err != nil {
+		return nil, xerrors.Errorf("render subject: %w", err)
+	}
+
+	htmlBody, err := render.HTML(bodyTmpl)
+	if err != nil {
+		return nil, xerrors.Errorf("render HTML body: %w", err)
+	}
+
+	plainBody, err := render.Plaintext(bodyTmpl)
+	if err != nil {
+		return nil, xerrors.Errorf("render plaintext body: %w", err)
+	}
+
+	// Then, reuse these strings in the HTML & plain body templates.
+	payload.Labels.Set("_subject", subject)
+	payload.Labels.Set("_body", htmlBody)
+	htmlBody, err = render.GoTemplate(htmlTemplate, payload, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("render full html template: %w", err)
+	}
+	payload.Labels.Set("_body", plainBody)
+	plainBody, err = render.GoTemplate(plainTemplate, payload, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("render full plaintext template: %w", err)
+	}
+
+	return s.dispatch(subject, htmlBody, plainBody, payload.UserEmail), nil
+}
+
+// dispatch returns a DeliveryFunc capable of delivering a notification via SMTP.
+//
+// NOTE: this is heavily inspired by Alertmanager's email notifier:
+// https://github.com/prometheus/alertmanager/blob/342f6a599ce16c138663f18ed0b880e777c3017d/notify/email/email.go
+func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) DeliveryFunc {
+	return func(ctx context.Context, msgID uuid.UUID) (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		default:
+		}
+
+		var (
+			c    *smtp.Client
+			conn net.Conn
+			err  error
+		)
+
+		s.log.Debug(ctx, "dispatching via SMTP", slog.F("msg_id", msgID))
+
+		// Dial the smarthost to establish a connection.
+		smarthost, smarthostPort, err := s.smarthost()
+		if err != nil {
+			return false, xerrors.Errorf("'smarthost' validation: %w", err)
+		}
+		if smarthostPort == "465" {
+			return false, xerrors.New("TLS is not currently supported")
+		}
+
+		var d net.Dialer
+		// Outer context has a deadline (see CODER_NOTIFICATIONS_DISPATCH_TIMEOUT).
+		conn, err = d.DialContext(ctx, "tcp", fmt.Sprintf("%s:%s", smarthost, smarthostPort))
```

Comment:
Yes it does. This is how Alertmanager does it currently.
We could establish multiple connections and reuse them if SMTP supports it, which I suspect it would.